### PR TITLE
Add more preg wrapper methods

### PR DESCRIPTION
**Pull request by johnstevenson**
I appreciate that you are already working on this, but since I was already testing various API ideas it seemed quicker to knock up a POC to show how this library could provide safe wrappers for the full range of preg_* function usage (plus its own API), rather than discussing it all. So no worries if you don't want this.

- Union types in the preg_* functions are supported (even though Composer may not use them itself)
- The $subject param from `PcreException::fromFunction` has been removed because it was not used
- `ReplaceResult::result` is now a `string|string[]`
- `Regex::isMatch` has been introduced as syntax sugar for `Regex::match(...)->matched`

Tests are needed for:
- `Preg::filter`
- `Preg::grep`
- `Preg::replaceCallbackArray`
- `Preg::split`

There is an inconsistency between the preg_ functions that emit warnings (in that `Preg::filter` is the only one to suppress them) because I don't know what is best to do.

### CODE REVIEW

File: tests/RegexTest.php [hunk 1]
```diff
@@ -164,7 +164,7 @@ private function doExpectException($class, $message = null)
     private function doExpectWarning($message)
     {
         if (method_exists($this, 'expectWarning')) {
-            $this->expectWarning($message);
+            $this->expectWarning();
```
Comment (Seldaek) [inline tests/RegexTest.php:167]: Why this change?

File: src/PcreException.php [hunk 2]
```diff
@@ -15,14 +15,17 @@ class PcreException extends \RuntimeException
...
+     * @param  string|string[] $pattern
      * @return self
      */
-    public static function fromFunction($function, $pattern, $subject)
+    public static function fromFunction($function, $pattern)
     {
         $code = preg_last_error();
 
+        if (is_array($pattern)) {
+            $pattern = implode(',', $pattern);
```
Comment (Seldaek) [inline src/PcreException.php:26]:
```suggestion
            $pattern = implode(', ', $pattern);
```

**Review by Seldaek — COMMENTED** [review]
Overall I'm happy to move the logic to Preg and then use that and wrap in Result from Regex, that makes more sense.

I still disagree about supporting array inputs though, because I frankly don't remember ever needing this, and if you do you can as well foreach yourself, or we can add an array specific variant of the method returning always an array.

Because right now with stuff like `string|string[]` you end up with this kind of problem, which requires more type checks than strictly necessary, and it sucks IMO when you have to add code checks to please PHPStan even though you know it'll be fine, because the type system at the library level was poorly designed and is a mess 

See https://phpstan.org/r/5bc76d20-eb29-4d97-9e8b-fd7252a62a79

File: tests/RegexTest.php (hunk 1 above)
Comment (johnstevenson) [reply tests/RegexTest.php:167]: PhpStan picked up that `expectWarning` doesn't take any params: https://github.com/sebastianbergmann/phpunit/blob/ee1820c53e849c5da75f26ef4c09995a4ff457cc/src/Framework/TestCase.php#L445-L448

`expectWarningMessage` does though, so I'll change it to this.

File: src/PcreException.php (hunk 2 above)
Comment (johnstevenson) [reply src/PcreException.php:26]: Ok, will change.

**Comment by johnstevenson** [comment]
Thanks for illustrating the `string|string[]` issue. That's a very good point, and I get it now.

>I still disagree about supporting array inputs though

I presume you mean array $subjects? So we can:

1. Remove them completely (from the `Preg` class)
2. Allow them in the `Preg` class but remove them from the `Regex` class.
3. Allow them and add a `ReplaceArrayResult` class (naming?)

And your choice is?

File: tests/RegexTest.php (hunk 1 above)
Comment (johnstevenson) [reply tests/RegexTest.php:167]: Actually you need both:
```php
if (method_exists($this, 'expectWarning')) {
    $this->expectWarning();
    $this->expectWarningMessage($message);
}
```

File: src/Preg.php [hunk 3]
```diff
@@ -14,67 +14,164 @@
...
      */
-    public static function match($pattern, $subject, &$matches = null, $flags = 0, $offset = 0)
+    public static function filter($pattern, $replacement, $subject, $limit, &$count)
     {
-        $result = Regex::match($pattern, $subject, $flags, $offset);
-        $matches = $result->matches;
+        $error = null;
+        set_error_handler(function ($errno, $errstr) use (&$error) {
+            $error = $errstr;
+            return true;
```
[the remark this answers is not in the document]
Comment (Seldaek) [reply src/Preg.php]: Probably the most transparent would be to return false and call the previous error handler if any. Essentially we would just hook in to know if a warning happens or not

**Comment by Seldaek** [comment]
I'd remove them from both classes I think, to keep things simple for now, and add a dedicated method to both classes with dedicated result class later if we really need this?

File: src/Preg.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (johnstevenson) [reply src/Preg.php]: I'm leaning towards completely removing array $subjects (it makes things too complex), which in this case means that we can simply drop the error_handler, because `null` is returned.

**Comment by Seldaek** [comment]
> oh great namer of things

🙄 

Message seems fine to me :p

Thanks for the updates, let me review the latest state again.

File: tests/PregTest.php [hunk 4]
```diff
@@ -50,7 +50,7 @@ public function testMatchNoMatch()
     public function testMatchThrowsWithBadPatternIfWarningsAreNotThrowing()
     {
         $this->doExpectException('Composer\Pcre\PcreException', 'preg_match(): failed executing "{abc": '.(PHP_VERSION_ID >= 80000 ? 'Internal error' : (PHP_VERSION_ID >= 70201 ? 'PREG_INTERNAL_ERROR' : '' /* Ignoring here, some old versions return UNDEFINED_ERROR while some have been fixed */)));
-        @Preg::match('{abc', 'abcdefghijklmnopqrstuvwxyz');
+        @Preg::match('{abc', 'abcdefghijklmnopqrstuvwxyz', $matches);
```
Comment (Seldaek) [inline tests/PregTest.php:53]: Here I wouldn't add $matches either, or is there a reason you added it?

**Comment by Seldaek** [comment]
OK looks good to me, just the couple weird changes in test files which I don't understand, and perhaps adding grep/filter/etc. to the Regex class would be nice, if you have time, otherwise I can do some other time.

File: tests/PregTest.php (hunk 4 above)
Comment (johnstevenson) [reply tests/PregTest.php:53]: Yes, Preg::match requires at least 3 arguments, otherwise

```1) Composer\Pcre\PregTest::testMatchThrowsWithBadPatternIfWarningsAreNotThrowing
Failed asserting that exception of type "ArgumentCountError" matches expected exception "Composer\Pcre\PcreException". Message was: "Too few arguments to function Composer\Pcre\Preg::match(), 2 passed in C:\Users\John\projects\pcre\tests\PregTest.php on line 53 and at least 3 expected" at
C:\Users\John\projects\pcre\src\Preg.php:26
C:\Users\John\projects\pcre\tests\PregTest.php:53
```

File: tests/PregTest.php [hunk 5]
```diff
@@ -30,7 +30,7 @@ public function testMatch()
      */
     public function testMatchNoRef()
     {
-        $count = Preg::match('{(?P<m>[io])}', 'abcdefghijklmnopqrstuvwxyz');
+        $count = Preg::match('{(?P<m>[io])}', 'abcdefghijklmnopqrstuvwxyz', $matches);
```
[the remark this answers is not in the document]
Comment (johnstevenson) [reply tests/PregTest.php:33]: Which earlier comment? I must admit I couldn't quickly see what the NoRef tests were doing, but some of them errored (Preg::match requires the by-ref $matches param), so I fixed them and moved on with what I was doing at the time.

I'll remove them all.

File: tests/PregTest.php (hunk 4 above)
Comment (Seldaek) [reply tests/PregTest.php:44]: I think this fails because you made $matches non nullable in `Preg::match`, but https://www.php.net/preg_match has `$matches` as `&$matches=null` so that it is optional.

File: tests/PregTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (Seldaek) [reply tests/PregTest.php]: The earlier comment was "The only downside for users would be having to add the optional params if they are not being used." in https://github.com/composer/pcre/issues/2#issue-1068670371 - maybe I misunderstood.

[the remark this answers is not in the document]
Comment (johnstevenson) [reply tests/PregTest.php]: Ah, sorry I was wrong there

File: tests/PregTest.php (hunk 4 above)
Comment (johnstevenson) [reply tests/PregTest.php:44]: Yes, I've got confused somewhere along the way! I'll make the by-refs nullable

**Comment by Seldaek** [comment]
Thanks for the updates. Do you still want to add anything? Otherwise I'd merge this already. We're getting closer to something stable IMO.

**Comment by johnstevenson** [comment]
If you can hang on. I'm just adding missing tests now and I need to fix `preg_filter` which I hadn't realized returns null when the pattern doesn't match a string:

```php
<?php
preg_filter('/d/', 'e', 'abcd') // abce
preg_filter('/d/', 'e', 'abc')  // null
``` 

I'm thinking of allowing Preg::filter to accept arrays, because then it is a bit like preg_grep with a filter, but not adding it to the Regex class. What do you think?

**Comment by Seldaek** [comment]
But then you still have to return `string[]|string`? I'd say if anything I'd allow it *only* with array because if you're doing it on a single string there's really not a lot of benefit over `if(match) {replace}`.. Overall this function seems kinda weird to me I don't quite get the use case. I think it mostly is misused by people instead of preg_grep, if I can trust the comments on https://www.php.net/preg_filter

So maybe we just drop filter altogether? How about that? :p

**Comment by johnstevenson** [comment]
Yes I agree. Reading through the source to see how it worked, it turned out there wasn't even a test for it until last year!

**Comment by johnstevenson** [comment]
I think I'm done now. Please review, thanks.

**Comment by johnstevenson** [comment]
Sorry, just had to sort the tests out - they were so hard to read because they were all lumped together and it was easy to make (and not spot) mistakes.

I've only included pcre engine error (backtrack limit exhausted) tests for preg_match. Is it necessary to include them for the other functions?

**Comment by Seldaek** [comment]
> I've only included pcre engine error (backtrack limit exhausted) tests for preg_match. Is it necessary to include them for the other functions?

Nah that's fine by me. 

Thanks for all the test updates, looks good to me 👍🏻
